packages/camera/camera/ios/Classes/FLTThreadSafeTextureRegistry.m

```diff
@@ -0,0 +1,53 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "FLTThreadSafeTextureRegistry.h"
+
+@implementation FLTThreadSafeTextureRegistry {
+  NSObject<FlutterTextureRegistry> *_registry;
+}
+
+- (instancetype)initWithTextureRegistry:(NSObject<FlutterTextureRegistry> *)registry {
+  self = [super init];
+  if (self) {
+    _registry = registry;
+  }
+  return self;
+}
+
+- (int64_t)registerTexture:(NSObject<FlutterTexture> *)texture {
+  if (!NSThread.isMainThread) {
+    __block int64_t textureId;
+    // Use dispatch_sync to keep FlutterTextureRegistry's sychronous API, so that we don't introduce new potential race conditions.
+    // We do not break priority inversion here since it's the background thread waiting for main thread.
+    dispatch_sync(dispatch_get_main_queue(), ^{
```

Comment:
dispatch_sync made this method wait for main thread. It is obvious to the caller.
Maybe we introduce a completion handler and use dispatch_async instead so the wait is more obivous?

Reply:
I guess what I wanted to say in my comment is that, if we do not wait for main thread to complete (use `dispatch_async`), we effectively change the semantic of the wrapped function (from a sync API to an async one), which may introduce a potential race condition. But looks like my comment was not clearly phrased.

Reply:
Just in case miscommunication, consider this dummy code: 

```
var _foo = 0
func setFooTo2() {
  _foo = 2
}
func whatever() {
  setFooTo2()
  // since `setFooTo2()` is synchronous API, the subsequent code can safely assume `_foo` is updated
  let bar = 1 / foo
}
```

However, if we create a wrapper and change the API to asynchronous one, we can't make that assumption anymore: 
```
func whatever() {
  setFooTo2Asynchronously() // async API
  // race condition here
  let bar = 1 / foo
}
```

So back to our example, after calling this texture API, there could be a potential code execution assuming that this API is already taken effect (this code execution may or may not be the code immediately following this texture API). Since we have like 10+ registered methods, there could be exponential amount of combination that we need to check for race, which can easily get out of hand. That's why it's important to keep the sync API here, and changing it to async may introduce new hard-to-find races.

Reply:
I think your comment is fine. 
 
The race condition could happen on `textureId`. if we use a completion block, the `textureId` would only be available inside the block. So the caller should be aware that the `textureId` is only accessible after the block is executed.

Your point makes sense too, the caller in this plugin always need to wait for the textureId to continue anyway. Maybe we can rename the method to make it clearer to the caller that this method will block the current thread and wait on main thread? Maybe something like: 

```objc
/**
 * Registers a `FlutterTexture` for usage in Flutter and returns an id that can be used to reference
 * that texture when calling into Flutter with channels. Textures must be registered on the
 * platform thread. On success returns the pointer to the registered texture, else returns 0.
 *
 * The registration happens on Main Thread. If the calling thread is not Main Thread, the current thread will be blocked until the registration is successful.
 */
- (int64_t)registerTextureSync:(NSObject<FlutterTexture> *)texture;

```

Reply:
> if we use a completion block, the textureId would only be available inside the block. So the caller should be aware that the textureId is only accessible after the block is executed.

We can use completion block for the code in the same function for sure, but we can't use it for separate functions. (e.g. `dispose` and `create` functions are separate). The issue here is that, it's hard to tell if the developers had made implicit assumption of the synchronous nature when implementing this plugin. If we were to rewrite this plugin from the beginning, I would be fine with either approaches. 

I am cool with the new name as you suggested.

Reply:
I understand with the `create` `dispose` example the data should be sent before `create` returns, but the caller in `handleMethodCallAsync` can make it synchronous in that case and wait to return in the completion block, instead of enforcing its synchronicity in the `-registerTexture:` API.